### PR TITLE
Update `load_ard` to support Sentinel-2 Collection 3 data, including `s2cloudless` cloud masking

### PR DESCRIPTION
**Pull request by robbibt**
### Proposed changes
This pull request implements a major update to the `load_ard` function to allow the function to support the new Sentinel-2 Collection 3 products. In particular:

- Support loading `ga_s2am_ard_3` and `ga_s2bm_ard_3` products
- Support masking using both `Fmask` and `s2cloudless` cloud masks - `Fmask` remains the default mask, but can be easily changed to `s2cloudless` by passing `mask_pixel_quality='s2cloudless'`
- Updated the `Using_load_ard.ipynb` notebook to document all new features
- Update tests to run on updated notebook
- Improved function docstrings to improve readability

I've also made two related changes:

- Deprecated the `predicate=` param for filtering data as this functionality is now supported by `datacube-core` itself via `dataset_predicate`
- Swapped the `mask_filters` cloud mask buffering code to be more intuitive: 'dilation' will now expand cloudy pixels, rather than shrink them. This is a backwards-incompatible change, so I've added a warning so that anyone using this functionality will know it's changed.

### Checklist (replace `[ ]` with `[x]` to check off)
- [x] Notebook created using the [DEA-notebooks template](https://github.com/GeoscienceAustralia/dea-notebooks/tree/develop)
- [x] Remove any unused Python packages from `Load packages`
- [x] Remove any unused/empty code cells
- [x] Remove any guidance cells (e.g. `General advice`)
- [x] Ensure that all code cells follow the [PEP8 standard](https://www.python.org/dev/peps/pep-0008/) for code. The `jupyterlab_code_formatter` tool can be used to format code cells to a consistent style: select each code cell, then click `Edit` and then one of the `Apply X Formatter` options (`YAPF` or `Black` are recommended).
- [x] Include relevant tags in the final notebook cell (refer to the [DEA Tags Index](https://docs.dea.ga.gov.au/genindex.html), and re-use tags if possible)
- [x] Clear all outputs, run notebook from start to finish, and save the notebook in the state where all cells have been sequentially evaluated
- [ ] Test notebook on both the `NCI` and `DEA Sandbox` (flag if not working as part of PR and ask for help to solve if needed)
- [ ] If applicable, update the `Notebook currently compatible with the NCI|DEA Sandbox environment only` line below the notebook title to reflect the environments the notebook is compatible with

### CODE REVIEW

**Review by uchchwhash — APPROVED** [review]

File: Tools/dea_tools/datahandling.py [hunk 1]
```diff
@@ -91,107 +94,128 @@ def _common_bands(dc, products):
...
     predicate : function, optional
-        An optional function that can be passed in to restrict the
-        datasets that are loaded by the function. A predicate function
-        should take a `datacube.model.Dataset` object as an input (i.e.
-        as returned from `dc.find_datasets`), and return a boolean.
-        For example, a predicate function could be used to return True
-        for only datasets acquired in January:
-        `dataset.time.begin.month == 1`
+        DEPRECATED: Please use `dataset_predicate` instead.
+        An function that can be passed in to restrict the datasets that
```
Comment (geoscience-aman) [inline Tools/dea_tools/datahandling.py]: Missing word I think. Should be "An optional function".

File: Tools/dea_tools/datahandling.py [hunk 2]
```diff
@@ -205,9 +229,9 @@ def load_ard(dc,
 
     Returns
     -------
-    combined_ds : xarray Dataset
-        An xarray dataset containing only satellite observations that
-        contains greater than `min_gooddata` proportion of good quality
+    combined_ds : xarray.Dataset
+        An xarray.Dataset containing only satellite observations that
+        contain greater than `min_gooddata` proportion of good quality
```
Comment (geoscience-aman) [inline Tools/dea_tools/datahandling.py]: Perhaps reads better as "An xarray.Dataset containing only satellite observations with a proportion of good quality pixels greater than 'min_gooddata'."

**Review by geoscience-aman — APPROVED** [review]
This is a super awesome update @robbibt! :clap:

As part of my testing I: 

- [x] Checked that the output of [Frequently_used_code/Using_load_ard.ipynb ](https://github.com/GeoscienceAustralia/dea-notebooks/blob/develop/Frequently_used_code/Using_load_ard.ipynb) is working as expected. 
- [x] Checked that the new `load_ard()` function with `s2cloudless` masking produces the same output as manual `s2cloudless` masking with `dc.load()` in DEA_datasets/DEA_Sentinel2_Surface_Reflectance.ipynb (PR #983). 
- [x] Verified that the `min_gooddata` argument inside `load_ard()` was doing the cuts as expected. 

My comments are minor and mostly relate to making the user experience clearer in the [Frequently_used_code/Using_load_ard.ipynb ](https://github.com/GeoscienceAustralia/dea-notebooks/blob/develop/Frequently_used_code/Using_load_ard.ipynb) notebook.

Please let me know if you have any issues or clarifications. Once again, awesome job! 🎉

**Comment by geoscience-aman** [comment]
Ugh sorry for the emails everyone. Not sure why ReviewNB decided to post each comment separately. 😅

**Review by supermarkion — APPROVED** [review]
Thanks @robbibt about the changes. It is wonderful to have a ARD loading section here, which simplify downstream product loading/clean-up task.

**Comment by erialC-P** [comment]
<p>For simplicity, I have changed the index here to plot image 0 for continuity with the 'False' cloud masking cell immediately below. This also gets around the issue of the swath edge related data-absences.</p>

---

 <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by erialC-P** [comment]
<p>Agreed! Robbi has provided some text to address this.</p>

---

 <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by erialC-P** [comment]
<p>I've edited the notebook markdown here for clarity. </p><p>If we know of/can find a good resource explaining the difference between the cloud masks then I think we can add that later. For now, the decision is likely to be made based on user testing and whether the user dataset is sensor specific or blended.</p>

---

 <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by robbibt** [comment]
> View / edit / reply to [this conversation](https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/) on ReviewNB
> 
> @geoscience-aman commented on 2023-01-11T02:37:26Z_ ----------------------------------------------------------------
> 
> Is there a way around having to run `load_ard()` again just to access the `s2cloudless_prob` layer? I find it slightly weird that the probability layer is masked if `mask_pixel_quality=False` is not specified.

Yeah, this is a bit weird to me too... I was thinking about adding some magic code under the hood that excludes `s2cloudless_prob` from being masked, but wasn't sure if treating those layers differently from the others would produce unexpected confusion, e.g. could there be some unanticipated use case for cloud masking that layer that we haven't thought of? Might flag this with a Github issue as something to think about and possibly implement outside of this PR

**Comment by erialC-P** [comment]
I was also thinking that this should be logged as an issue for later but that we could leverage it for now to compare the performance of Fmask and s2cloudless to support S2 decision making. Any aversions to me adding a cell to the notebook to this effect?

View / edit / reply to this conversation<https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/> on ReviewNB

geoscience-aman commented on 2023-01-11T02:37:26Z ----------------------------------------------------------------

Is there a way around having to run load_ard() again just to access the s2cloudless_prob layer? I find it slightly weird that the probability layer is masked if mask_pixel_quality=False is not specified.

Yeah, this is a bit weird to me too... I was thinking about adding some magic code under the hood that excludes s2cloudless_prob from being masked, but wasn't sure if treating those layers differently from the others would produce unexpected confusion, e.g. could there be some unanticipated use case for cloud masking that layer that we haven't thought of? Might flag this with a Github issue as something to think about and possibly implement outside of this PR

**Comment by erialC-P** [comment]
<p>Interesting...you could definitely get tripped up by that if you forgot that Fmask were applied by default. </p>

---

 <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by erialC-P** [comment]
<p>I don't understand the <code>dc_landsat</code> syntax here either. On an assumption that it's meant to be <code>dc</code>, I've changed it in my reviewed copy of the notebook.  </p>

---

 <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by erialC-P** [comment]
<p>I've removed the <em>explicit</em> and <em>query</em> hyperlinks in my reviewed notebook and replaced them with italics instead.</p>

---

 <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by erialC-P** [comment]
<p>I think it's ok in this instance because we're not loading fragments of the total dataset but rather loading everything. The notebook also links out to the <code>Parallel_processing_with_Dask</code> notebook so users can redirect there to clarify any confusion. </p>

---

 <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: Tools/dea_tools/datahandling.py (hunk 1 above)
Comment (erialC-P) [reply Tools/dea_tools/datahandling.py]: revised

File: Tools/dea_tools/datahandling.py (hunk 2 above)
Comment (erialC-P) [reply Tools/dea_tools/datahandling.py]: revised

**Review by erialC-P — APPROVED** [review]
Looks great! Thanks for updating all the new load_ard options in this notebook.
I have made a few small grammatical changes and added a couple of explainers into the notebook for completeness but no major changes.

**Comment by robbibt** [comment]
> View / edit / reply to [this conversation](https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/) on ReviewNB
> 
> _erialC-P commented on 2023-01-11T05:35:02Z_ ----------------------------------------------------------------
> 
> Unusual error returned following "Loading 12 time steps". Doesn't appear to affect the analysis. 
> ```
> CPLReleaseMutex: Error = 1 (Operation not permitted)
> ```

This is annoying but there's nothing we can do - I've asked about it many times but there's no clear workaround :(

**Comment by robbibt** [comment]
<p>Awesome, I think that sounds like an easy solution for now!</p>

---

 <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/984/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Pull request merged**